atc/exec/set_pipeline_step.go

```diff
@@ -0,0 +1,244 @@
+		spParams.PipelineName = step.metadata.PipelineName
+	}
+
+	team, found, err := step.delegate.FindTeam()
+	if err != nil {
+		logger.Error("failed-to-find-team", err)
+		return err
+	}
+	if !found {
+		err = fmt.Errorf("team not found")
+		logger.Error("failed-to-find-team", err)
+		return err
+	}
+
+	fromVersion := db.ConfigVersion(0)
+	pipeline, found, err := team.Pipeline(spParams.PipelineName)
+	if err != nil {
+		return nil
+	}
+
+	var existingConfig atc.Config
+	if !found {
+		existingConfig = atc.Config{}
+	} else {
+		fromVersion = pipeline.ConfigVersion()
+		existingConfig, err = pipeline.Config()
+		if err != nil {
+			return err
+		}
+	}
+
+	diffExists := existingConfig.Diff(stdout, *atcConfig)
+	if !diffExists {
+		logger.Debug("no-diff")
+
+		if spParams.FailWhenNoDiff {
+			return errors.New("no diff found")
+		}
+
+		fmt.Fprintf(stdout, "No diff found.\n")
+		step.succeeded = true
+		step.delegate.Finished(logger, ExitStatus(0), fromVersion, fromVersion)
+		return nil
+	}
+
+	fmt.Fprintf(stdout, "Updating the pipeline.\n")
+	pipeline, _, err = team.SavePipeline(spParams.PipelineName, *atcConfig, fromVersion, false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(stdout, "Done successfully.\n")
+	logger.Info("saved-pipeline", lager.Data{"team": team.Name(), "pipeline": pipeline.Name()})
+	step.succeeded = true
+	step.delegate.Finished(logger, ExitStatus(0), fromVersion, pipeline.ConfigVersion())
+
+	return nil
+}
+
+func (step *SetPipelineStep) Succeeded() bool {
+	return step.succeeded
+}
+
+type setPipelineArtifactSource struct {
+	params atc.SetPipelineParams
+	ctx    context.Context
+	logger lager.Logger
+	repo   *artifact.Repository
+}
+
+// streamInBytes streams a file from other resource and returns a byte array.
+func (s setPipelineArtifactSource) streamInBytes(path string) ([]byte, error) {
+	segs := strings.SplitN(path, "/", 2)
+	if len(segs) != 2 {
+		return nil, UnspecifiedArtifactSourceError{path}
+	}
+
+	sourceName := artifact.Name(segs[0])
+	filePath := segs[1]
+
+	source, found := s.repo.SourceFor(sourceName)
+	if !found {
+		return nil, UnknownArtifactSourceError{sourceName, path}
+	}
+
+	stream, err := source.StreamFile(s.ctx, s.logger, filePath)
+	if err != nil {
+		if err == baggageclaim.ErrFileNotFound {
+			return nil, fmt.Errorf("task config '%s/%s' not found", sourceName, filePath)
+		}
+		return nil, err
+	}
+
+	defer stream.Close()
+
+	byteConfig, err := ioutil.ReadAll(stream)
+	if err != nil {
+		return nil, err
+	}
+
+	return byteConfig, nil
+}
+
+// FetchConfig streams pipeline configure file and var files from other resources
+// and construct an atc.Config object.
+func (s setPipelineArtifactSource) FetchConfig() (*atc.Config, error) {
+	config, err := s.streamInBytes(s.params.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	staticVarss := []vars.Variables{}
+	if len(s.params.Var) > 0 {
+		staticVarss = append(staticVarss, vars.StaticVariables(s.params.Var))
```

Comment:
In my fork I planned to use the step-level `vars:` config:

https://github.com/concourse/concourse/compare/feature/set-pipeline-step#diff-cd1e4bcee1dd2a9f163ac91a8232bcdeR327

There's no top-level `load_vars_from` at the moment but it could be added.

Reply:
I have moved `vars` to step-level and added `var_files`. To our use case, `var_files` should be mandatory.